Resolve configuration errors

debelistic opened the pull request:
### What does this PR do?

- set sequelizerc for transpiling to native JS
- set sequelize scripts
- set development environment executable
- seed admin user to database using environment secrete
- remove production configuration errors

### Description of Task to be completed?

- add @babel/register to sequelize configuration
- use coveralls badge in README
- add scripts for database migrations and seeders
- setup tasks running in development environment
- remove devdependencies from production executable file
- add admin status to User model
- add admin seeder

### How should this be manually tested?

- run `npm run reset-db` to drop and migrate database
- run `npm run seeders` to add admin, ensure environment secrete is set in `.env` file
- run `npm run devstart` to see app in development environment
- run `npm start` to see app in production environment

### Any background context you want to provide?

The `dev ` file is, so we can run development execute file independently of the `www` file. In the `www` file we are making use of the compiled files from our `dist` folder, while the `dev` file makes use of ES6 code from `src` folder, and we require babel dependencies at the beginning so it can compile our code. 

The reason for this is that;
-  the current `www` requires babel which is a devDependency, and app would crash in production since Heroku only install project dependencies
-  we could move the babel dependencies to production as we are doing presently, however compiling code in production is inefficient.

### What are the relevant pivotal tracker stories?
#168250646

### Implementation Plan

N/A

### Questions:

- suggest any other config errors or conventions to be resolved in your review

voke1 commented:
Good work @debelistic However, I got this error when I ran `npm run test` ```ERROR: no pg_hba.conf entry for host "129.56.74.94", user "uahtduhrsmfqxc", database "dfrvfn41s4u850", SSL off```. 

I had to set ssl to true by specifying this ```dialectOptions: {
      ssl: true,
    }``` in the `config.js` file under both test and development for it to work. You may want to check that.


I also get this when I ran `npm run seeders` 

``` 
ERROR: Validation error
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! express-authorshaven@1.0.0 seeders: `sequelize db:seed:all`
npm ERR! Exit status 1
``` 
I'm presuming this might have to do with the seeder file

debelistic commented:
> Good work @debelistic However, I got this error when I ran `npm run test` `ERROR: no pg_hba.conf entry for host "129.56.74.94", user "uahtduhrsmfqxc", database "dfrvfn41s4u850", SSL off`.
> 
> I had to set ssl to true by specifying this `dialectOptions: { ssl: true, }` in the `config.js` file under both test and development for it to work. You may want to check that.
> 
> I also get this when I ran `npm run seeders`
> 
> ```
> ERROR: Validation error
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1
> npm ERR! express-authorshaven@1.0.0 seeders: `sequelize db:seed:all`
> npm ERR! Exit status 1
> ```
> 
> I'm presuming this might have to do with the seeder file

The error is with your local postgres configuration

voke1 commented:
> > Good work @debelistic However, I got this error when I ran `npm run test` `ERROR: no pg_hba.conf entry for host "129.56.74.94", user "uahtduhrsmfqxc", database "dfrvfn41s4u850", SSL off`.
> > I had to set ssl to true by specifying this `dialectOptions: { ssl: true, }` in the `config.js` file under both test and development for it to work. You may want to check that.
> > I also get this when I ran `npm run seeders`
> > ```
> > ERROR: Validation error
> > npm ERR! code ELIFECYCLE
> > npm ERR! errno 1
> > npm ERR! express-authorshaven@1.0.0 seeders: `sequelize db:seed:all`
> > npm ERR! Exit status 1
> > ```
> > 
> > 
> > I'm presuming this might have to do with the seeder file
> 
> The error is with your local postgres configuration

which local postgres configuration? I worked directly on your branch :)

debelistic commented:
> > > Good work @debelistic However, I got this error when I ran `npm run test` `ERROR: no pg_hba.conf entry for host "129.56.74.94", user "uahtduhrsmfqxc", database "dfrvfn41s4u850", SSL off`.
> > > I had to set ssl to true by specifying this `dialectOptions: { ssl: true, }` in the `config.js` file under both test and development for it to work. You may want to check that.
> > > I also get this when I ran `npm run seeders`
> > > ```
> > > ERROR: Validation error
> > > npm ERR! code ELIFECYCLE
> > > npm ERR! errno 1
> > > npm ERR! express-authorshaven@1.0.0 seeders: `sequelize db:seed:all`
> > > npm ERR! Exit status 1
> > > ```
> > > 
> > > 
> > > I'm presuming this might have to do with the seeder file
> > 
> > 
> > The error is with your local postgres configuration
> 
> which local postgres configuration? I worked directly on your branch :)

`pg_hba.conf` should be a file or plugin for postgres configuration

misanj commented on src/bin/dev:
```diff
@@ -0,0 +1,13 @@
+require('@babel/core');
+const http = require('http');
+const debug = require('debug');
+const app = require('../index').default;
```
Good job @debelistic  why do we have a dev file in the bin folder? It contains the exact same block of code in the www file except for line 5 that differs in path.

anneKay replied:
Please don't resolve reviews without attending to them or providing some answers to the questions

misanj reviewed (approved):
I now Understand what you where trying to do with the `dev` file and the `www` file after running the build script. for the `pg_hba.conf` error, the fix is to require postgres ssl in the `.env` file like this `PGSSLMODE=require`

I have tested this branch and it works fine.... well done @debelistic

voke1 commented:
**debelistic** 👍  works fine now

victor-nach commented on .envSample:
```diff
@@ -1,9 +1,11 @@
+TEST_DB_NAME=
+TEST_DB_USER=
+TEST_DB_PASSWORD=
+TEST_DB_HOST=
```
add a new line here.

anneKay replied:
Please add a new line here

debelistic replied:
I just did

victor-nach commented on package.json:
```diff
@@ -5,17 +5,17 @@
+    "devstart": "set DEBUG=dev && babel-watch ./src/bin/dev",
     "build": "rm -rf dist && mkdir dist && babel src -s -d dist",
-    "test": "set NODE_ENV=test&& npm run reset:db && mocha -r @babel/register --recursive src/tests/ --exit --timeout 10000",
+    "test": " set NODE_ENV=test&& npm run reset-db&& mocha -r @babel/register --recursive src/tests/ --exit",
```
why did you remove the '--timeout 10000', due to the default timeout being too small, some of our tests would fail.

debelistic replied:
It will encourage longer response, if there's any response that's taking long, we should refactor. Besides 10s is long for response and I tested without the timeout and everything works fine

anneKay replied:
Any reason why you removed the timeout?

anneKay replied:
When we finally have several tests setup and running, it might timeout and cause tests to fail, adding this timeout limit doesn't necessarily mean that the tests must stay this duration before completing, please add this back.

victor-nach commented on src/database/migrations/20190827125917-create-user.js:
```diff
@@ -13,7 +13,7 @@ module.exports = {
       },
     },
-    userName: {
+    username: {
```
I don't think you should remove the camel case here

anneKay replied:
Any reason for this? If you would like to update columns names, that should be done with another migration file.

victor-nach commented on src/models/user.js:
```diff
@@ -1,6 +1,6 @@
 export default (sequelize, DataTypes) => {
   const User = sequelize.define('User', {
-    userName: {
+    username: {
```
same here too, with the camel case

victor-nach commented on src/tests/routes/index.spec.js:
```diff
@@ -1,6 +1,6 @@
 import chai from 'chai';
 import chaiHttp from 'chai-http';
-import app from '..';
+import app from '../..';
```
why did you move the test file insde the tests / routes folder, do you intend for us to write tests separately for our controllers and models?

debelistic replied:
That was how Anna specified it when set up the test, I was going to run it by you before I made the changes. I'm guess we'll be testing models and controllers, maybe

debelistic replied:
That was the intruction from ttl, when I setup test

victor-nach commented on src/tests/routes/index.spec.js:
```diff
@@ -22,7 +23,8 @@ describe('GET /', () => {
     chai.request(app)
       .get(`${endPoint}`)
-      .end((_err, res) => {
+      .end((err, res) => {
```
if you're removing the _ from 'err', eslint would flag that, so you would also have to go and add a rule exception in the eslintrc file, or just stick with adding the underscore for unused variables.

debelistic replied:
I added an if statement that makes use of the error, such that it would return err if there's an error with the request

debelistic committed:
chore(config) - resolve configuration errors

- move seeders and migrations to database folder
- rename config path for seeders and migrations
- specify db config path for specific environments
- move routes to test routes test folders
- add @babel/register to sequelize configuration
- use coveralls badge in README
- add scripts for database migrations and seeders
- setup tasks running in development environment
- remove devdependencies from production executable file
- add admin status to User model
- add admin seeder
- remove username from test and validations
- specify db_url for database configuration

[Finishes #168250646]

anneKay commented:
Please don't resolve conversations without attending to the comments or providing answers/clarity to the questions

anneKay commented on src/validation/userValidation.js:
```diff
@@ -5,9 +5,9 @@ const userValidation = [
+  check('username').trim().not().isEmpty()
     .withMessage('username field cannot be empty.'),
-  check('userName').isAlphanumeric().withMessage('username should contain only letters and numbers'),
+  check('username').isAlphanumeric().withMessage('username should contain only letters and numbers'),
```
Please return all these fields(userName) to what they were

debelistic replied:
We have dropped username field entirely from the user object

debelistic committed:
chore(config) - resolve configuration errors

- remove DB  vars from app.json

[Finishes #168250646]

victor-iheanacho commented on src/tests/routes/userSpec.js:
```diff
@@ -1,6 +1,6 @@
 import chai from 'chai';
 import chaiHttp from 'chai-http';
-import app from '..';
+import app from '../..';
```
why did you move the tests into the routes folder?

debelistic replied:
The tests for routes are meant to be in the folder, I think that's why ttl had us set the test folder up like that

debelistic commented:
> Please don't resolve conversations without attending to the comments or providing answers/clarity to the questions
I've provided answers to all questions on the PR
